cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AbstractAppLifecycleHttpHandler.java

```diff
@@ -0,0 +1,215 @@
+      String versionId) throws BadRequestException {
+    if (appId == null) {
+      throw new BadRequestException("Path parameter app-id cannot be empty");
+    }
+    if (!EntityId.isValidId(appId)) {
+      throw new BadRequestException(String.format("Invalid app name '%s'", appId));
+    }
+    if (versionId == null) {
+      throw new BadRequestException("Path parameter version-id cannot be empty");
+    }
+    if (EntityId.isValidVersionId(versionId)) {
+      return namespaceId.app(appId, versionId);
+    }
+    throw new BadRequestException(String.format("Invalid version '%s'", versionId));
+  }
+
+  protected NamespaceId validateNamespace(@Nullable String namespace)
+      throws BadRequestException, NamespaceNotFoundException, AccessException {
+
+    if (namespace == null) {
+      throw new BadRequestException("Path parameter namespace-id cannot be empty");
+    }
+
+    NamespaceId namespaceId;
+    try {
+      namespaceId = new NamespaceId(namespace);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(String.format("Invalid namespace '%s'", namespace), e);
+    }
+
+    try {
+      if (!namespaceId.equals(NamespaceId.SYSTEM)) {
+        namespaceQueryAdmin.get(namespaceId);
+      }
+    } catch (NamespaceNotFoundException | AccessException e) {
+      throw e;
+    } catch (Exception e) {
+      // This can only happen when NamespaceAdmin uses HTTP calls to interact with namespaces.
+      // In AppFabricServer, NamespaceAdmin is bound to DefaultNamespaceAdmin, which interacts directly with the MDS.
+      // Hence, this exception will never be thrown
+      throw Throwables.propagate(e);
+    }
+    return namespaceId;
+  }
+
+  protected ProgramTerminator createProgramTerminator() {
+    return programId -> {
+      switch (programId.getType()) {
+        case SERVICE:
+        case WORKER:
+          killProgramIfRunning(programId);
+          break;
+        default:
+          break;
+      }
+    };
+  }
+
+  protected void killProgramIfRunning(ProgramId programId) {
+    ProgramRuntimeService.RuntimeInfo programRunInfo = findRuntimeInfo(programId, runtimeService);
+    if (programRunInfo != null) {
+      ProgramController controller = programRunInfo.getController();
+      controller.kill();
+    }
+  }
+
+  protected ApplicationRecord getApplicationRecord(ApplicationWithPrograms deployedApp) {
+    return new ApplicationRecord(
+        ArtifactSummary.from(deployedApp.getArtifactId().toApiArtifactId()),
+        deployedApp.getApplicationId().getApplication(),
+        deployedApp.getApplicationId().getVersion(),
+        deployedApp.getSpecification().getDescription(),
+        Optional.ofNullable(deployedApp.getOwnerPrincipal()).map(KerberosPrincipalId::getPrincipal)
+            .orElse(null),
+        deployedApp.getChangeDetail(), null);
+  }
+
+  protected BodyConsumer deployAppFromArtifact(
+      final ApplicationId appId,
+      final boolean skipMarkingLatest)
```

Comment:
@GnsP  This is the only change in the moved function right ?

Reply:
yes.